modules/dnf: Substitute variables in DNF cache path

BenoitKnecht opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
The cache directory can be specified with variables that are expanded by DNF, e.g.

```
cachedir=/var/cache/yum/$basearch/$releasever
```

But the `dnf` module would use that path literally, instead of replacing `$basearch` and `$releasever` with their values.

This commit ensures that variables in `cachedir` are properly substituted.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`dnf`

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

To reproduce the issue, set

```ini
[main]
cachedir=/var/cache/yum/$basearch/$releasever
```

in `/etc/dnf/dnf.conf` and use the `dnf` module to install a package. On the target system, Ansible will create a directory called `/var/cache/yum/$basearch/$releasever`, instead of `/var/cache/yum/x86_64/8` for instance.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

BenoitKnecht committed:
modules/dnf: Substitute variables in DNF cache path

The cache directory can be specified with variables that are expanded by DNF,
e.g.

```
cachedir=/var/cache/yum/$basearch/$releasever
```

But the `dnf` module would use that path literally, instead of replacing
`$basearch` and `$releasever` with their values.

This commit ensures that variables in `cachedir` are properly substituted.

Signed-off-by: Benoît Knecht <bknecht@protonmail.ch>

sivel commented on lib/ansible/modules/dnf.py:
```diff
@@ -577,6 +578,7 @@ def _ensure_dnf(self):
             import dnf.exceptions
             import dnf.subject
             import dnf.util
+            import libdnf
```
Up a few lines, you'll need to add a `global libdnf`

BenoitKnecht replied:
Thanks @sivel, added! :slightly_smiling_face:

BenoitKnecht commented:
Looks like [`mypy` is unhappy about `libdnf`](https://dev.azure.com/ansible/ansible/_build/results?buildId=69838&view=logs&j=f39c292d-0a23-56d4-e7e5-386951c14ece&t=3af4f6c8-91cf-5ddb-8873-9e1cb9dedca9&l=361) now though:

```
01:08 FATAL: Encountered 1 fatal errors reported by mypy:
01:08 lib/ansible/modules/dnf.py:582:1: import: Cannot find implementation or library stub for module named "libdnf"
```

I'm not sure how to get around that.

BenoitKnecht committed:
mypy: Ignore missing imports for libdnf

Now that the `dnf` module uses `libdnf`, `mypy` returns the following error on
`lib/ansible/modules/dnf.py`:

```
import: Cannot find implementation or library stub for module named "libdnf"
```

This commit makes `mypy` ignore missing imports for `libdnf`.

Signed-off-by: Benoît Knecht <bknecht@protonmail.ch>

BenoitKnecht commented:
> Looks like [`mypy` is unhappy about `libdnf`](https://dev.azure.com/ansible/ansible/_build/results?buildId=69838&view=logs&j=f39c292d-0a23-56d4-e7e5-386951c14ece&t=3af4f6c8-91cf-5ddb-8873-9e1cb9dedca9&l=361) now though:
> 
> ```
> 01:08 FATAL: Encountered 1 fatal errors reported by mypy:
> 01:08 lib/ansible/modules/dnf.py:582:1: import: Cannot find implementation or library stub for module named "libdnf"
> ```
> 
> I'm not sure how to get around that.

OK, I managed to work around it with 197e1a6771a16d1ec8459ed96c5d16a1d13f0eee.

jborean93 reviewed (changes requested):
For the changes here we will need a changelog fragment for this PR which is documented here https://docs.ansible.com/ansible/latest/community/development_process.html#creating-changelog-fragments. We would also look at adding some integration tests for this scenario under https://github.com/ansible/ansible/tree/devel/test/integration/targets/dnf. The integration tests might be a bit harder to setup but we can certainly work through them if you have any questions.

Akasurde reviewed (approved):
LGTM
